File: frontend/src/lib/observation-attributes.ts
/**
 * What a person can record on a count row besides the number: sex, life
 * stage and behaviour. Mirrors `backend/app/core/observation_attributes.py`,
 * the same pairing as confidence.ts / confidence.py. The backend validates
 * against its copy, so a value that is not there is refused with a 422.
 *
 * Values are Camtrap DP's own (American spelling on the wire); the labels
 * are what the dropdowns show. Empty string in a select means unknown and
 * is sent as null.
 */

export interface AttributeOption {
  value: string;
  label: string;
}

export const SEX_OPTIONS: AttributeOption[] = [
  { value: "female", label: "Female" },
  { value: "male", label: "Male" },
];

export const LIFE_STAGE_OPTIONS: AttributeOption[] = [
  { value: "adult", label: "Adult" },
  { value: "subadult", label: "Subadult" },
  { value: "juvenile", label: "Juvenile" },
];

export const BEHAVIOR_OPTIONS: AttributeOption[] = [
  { value: "traveling", label: "Travelling" },
  { value: "foraging", label: "Foraging" },
  { value: "resting", label: "Resting" },
  { value: "vigilance", label: "Vigilance" },
  { value: "drinking", label: "Drinking" },
  { value: "grooming", label: "Grooming" },
  { value: "courtship", label: "Courtship" },
  { value: "nursing", label: "Nursing" },
  { value: "aggression", label: "Aggression" },
  { value: "marking", label: "Marking" },
];

/** The three dropdowns of a count row, in display order. */
export const OBSERVATION_ATTRIBUTES = [
  { field: "sex", label: "Sex", options: SEX_OPTIONS },
  // "Age" on screen: the panel is narrow and "Life stage" truncates. The
  // export column keeps Camtrap DP's `life_stage`.
  { field: "life_stage", label: "Age", options: LIFE_STAGE_OPTIONS },
  { field: "behavior", label: "Behaviour", options: BEHAVIOR_OPTIONS },
] as const;

export type ObservationAttributeField =
  (typeof OBSERVATION_ATTRIBUTES)[number]["field"];

File: backend/app/core/observation_attributes.py
"""The fixed vocabularies for what a person can record on an observation
cohort besides the count: sex, life stage and behaviour.

One place for the backend; `frontend/src/lib/observation-attributes.ts`
mirrors it for the dropdowns, the same pairing as confidence.py /
confidence.ts. The API validates against these, so the frontend copy can
never put a value in the database that this file does not know.

`sex` and `life_stage` are Camtrap DP's own enums, so they export as is.
`behavior` is free text in the standard ("preferably controlled values");
the list here is the one AddaxAI Connect uses, kept identical so the two
products' exports share a vocabulary. American spelling on the wire and
in the database, matching the standard's column name.

NULL means unknown everywhere. Never store a literal "unknown": it is not
in the Camtrap DP enums and every query would have to special-case it.
"""

from typing import Literal

SEXES = ("female", "male")
LIFE_STAGES = ("adult", "subadult", "juvenile")
BEHAVIORS = (
    "traveling",
    "foraging",
    "resting",
    "vigilance",
    "drinking",
    "grooming",
    "courtship",
    "nursing",
    "aggression",
    "marking",
)

Sex = Literal["female", "male"]
LifeStage = Literal["adult", "subadult", "juvenile"]
Behavior = Literal[
    "traveling",
    "foraging",
    "resting",
    "vigilance",
    "drinking",
    "grooming",
    "courtship",
    "nursing",
    "aggression",
    "marking",
]
